Source/CCZ4/ADMQuantities.hpp

```diff
@@ -0,0 +1,132 @@
+/* GRChombo
+ * Copyright 2012 The GRChombo collaboration.
+ * Please refer to LICENSE in GRChombo's root directory.
+ */
+
+#ifndef ADMQUANTITIES_HPP_
+#define ADMQUANTITIES_HPP_
+
+#include "ADMConformalVars.hpp"
+#include "CCZ4Geometry.hpp"
+#include "Cell.hpp"
+#include "Coordinates.hpp"
+#include "FourthOrderDerivatives.hpp"
+#include "GRInterval.hpp"
+#include "Tensor.hpp"
+#include "TensorAlgebra.hpp"
+#include "UserVariables.hpp" //This files needs NUM_VARS - total number of components
+#include "VarsTools.hpp"
+#include "simd.hpp"
+
+//! Calculates the ADM massa
+class ADMQuantities
+{
+    // Use the variable definition in ADMConformalVars - only require the key
+    // vars
+    template <class data_t> using Vars = ADMConformalVars::VarsNoGauge<data_t>;
+
+    template <class data_t>
+    using Diff1Vars = ADMConformalVars::Diff2VarsNoGauge<data_t>;
+
+  public:
+    enum DIR
+    {
+        X,
+        Y,
+        Z
+    };
+
+    ADMQuantities(const std::array<double, CH_SPACEDIM> &a_center, double a_dx,
+                  int a_c_Madm = -1, int a_c_Jadm = -1, double a_G_Newton = 1.0)
+        : m_deriv(a_dx), m_center(a_center), m_G_Newton(a_G_Newton),
+          m_c_Madm(a_c_Madm), m_c_Jadm(a_c_Jadm), m_dir(Z)
+    {
+    }
+
+    // in case user wants to change direction of spin calculation to something
+    // other than Z
+    void set_spin_dir(DIR spin_direction) { m_dir = spin_direction; }
+
+    template <class data_t> void compute(Cell<data_t> current_cell) const
+    {
+        // copy data from chombo gridpoint into local variables, and calc 1st
+        // derivs
+        const auto vars = current_cell.template load_vars<Vars>();
+        const auto d1 = m_deriv.template diff1<Diff1Vars>(current_cell);
+
+        using namespace TensorAlgebra;
+        const auto h_UU = compute_inverse_sym(vars.h);
+
+        // Surface element for integration
+        Coordinates<data_t> coords(current_cell, m_deriv.m_dx, m_center);
+        Tensor<1, data_t> x = {coords.x, coords.y, coords.z};
+        Tensor<1, data_t> dS_U = x;
+
+        data_t dS_norm = 0.;
+        FOR(i, j) { dS_norm += vars.h[i][j] / vars.chi * dS_U[i] * dS_U[j]; }
+        dS_norm = sqrt(dS_norm);
+        FOR(i) { dS_U[i] /= dS_norm; }
+
+        Tensor<1, data_t> dS_L;
+        FOR(i)
+        {
+            // dS_L[i] = dS_U[i];
+            dS_L[i] = 0.;
+            FOR(j) { dS_L[i] += vars.h[i][j] / vars.chi * dS_U[j]; }
+        }
+
+        if (m_c_Madm >= 0)
+        {
+            data_t Madm = 0.0;
+            FOR(i, j, k, l)
+            {
+                Madm += dS_L[i] / (16. * M_PI * m_G_Newton) /
+                        (vars.chi * sqrt(vars.chi)) * h_UU[j][k] * h_UU[i][l] *
+                        (vars.chi * (d1.h[l][k][j] - d1.h[j][k][l]) -
+                         (vars.h[l][k] * d1.chi[j] - vars.h[j][k] * d1.chi[l]));
```

Comment:
Maybe add a reference for these formulae as a comment e.g. Baumgarte and Shapiro p85 Eq. (3.128)?